sycl/test-e2e/KernelCompiler/sycl_link.cpp

```diff
@@ -0,0 +1,102 @@
+//==----------- sycl_link.cpp --- kernel_compiler extension tests ----------==//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+// REQUIRES: (opencl || level_zero)
+// REQUIRES: aspect-usm_shared_allocations
+
+// UNSUPPORTED: accelerator
+// UNSUPPORTED-INTENDED: while accelerator is AoT only, this cannot run there.
+
+// -- Test for a simple linking case with source files compiled from SYCL source
+// -- at runtime.
+
+// RUN: %{build} -o %t.out
+// RUN: %{run} %t.out
+// RUN: %{l0_leak_check} %{run} %t.out
+
+#include <sycl/detail/core.hpp>
+#include <sycl/kernel_bundle.hpp>
+#include <sycl/usm.hpp>
+
+namespace syclex = sycl::ext::oneapi::experimental;
+using source_kb = sycl::kernel_bundle<sycl::bundle_state::ext_oneapi_source>;
+using obj_kb = sycl::kernel_bundle<sycl::bundle_state::object>;
+using exe_kb = sycl::kernel_bundle<sycl::bundle_state::executable>;
+
+// TODO: remove SYCL_EXTERNAL from the kernel once it is no longer needed.
+auto constexpr SYCLImportSource = R"===(
+#include <sycl/sycl.hpp>
+
+SYCL_EXTERNAL void TestFunc(int *Ptr, int Size);
+
+// use extern "C" to avoid name mangling
+extern "C" SYCL_EXTERNAL SYCL_EXT_ONEAPI_FUNCTION_PROPERTY(
+    (sycl::ext::oneapi::experimental::single_task_kernel))
+void TestKernel(int *Ptr, int Size) {
+  TestFunc(Ptr, Size);
+}
+
+)===";
+
+auto constexpr SYCLExportSource = R"===(
+#include <sycl/sycl.hpp>
+
+SYCL_EXTERNAL void TestFunc(int *Ptr, int Size) {
+  for (size_t I = 0; I < Size; ++I)
+    Ptr[I] = I;
+}
+
+)===";
+
+int main() {
+  sycl::queue Q;
+
+  if (!Q.get_device().ext_oneapi_can_compile(syclex::source_language::sycl)) {
+    std::cout << "Device does not support one of the source languages: "
+              << Q.get_device().get_info<sycl::info::device::name>()
+              << std::endl;
+    return 0;
+  }
+
+  source_kb ImportSourceKB = syclex::create_kernel_bundle_from_source(
+      Q.get_context(), syclex::source_language::sycl, SYCLImportSource);
+  source_kb ExportSourceKB = syclex::create_kernel_bundle_from_source(
+      Q.get_context(), syclex::source_language::sycl, SYCLExportSource);
+
+  syclex::properties BuildOpts{
+      syclex::build_options{"-fsycl-allow-device-image-dependencies"}};
+  obj_kb ImportObjKB = syclex::compile(ImportSourceKB, BuildOpts);
+  obj_kb ExportObjKB = syclex::compile(ExportSourceKB, BuildOpts);
+
+  exe_kb ExecKB = sycl::link({ImportObjKB, ExportObjKB});
```

Comment:
Question/idea for an additional test: Compile another source bundle, also defining `TestFunc` with a different body, first, then link `{ImportObjKB, ExportObjKB}` -- is the correct version of the exported symbol picked up? (I fear this might clash in the lookup from `m_ExportSymbolImages`).

Reply:
Good shout! I've added a test like this and it seems to work as intended currently.

It also got me thinking: Kernel bundles using imported symbols shouldn't be able to resolve their symbols implicitly, so I added another test for that too, which seems to also work as intended out-of-the-box. However, it makes me wonder if we should explicitly prevent these kernel bundles from registering their exported symbols in `m_ExportedSymbolImages`, e.g. by adding an argument to `addImage` that skips this registration. Thoughts?

Reply:
> I've added a test like this and it seems to work as intended currently.

Nice, thanks. I had overlooked that `ProgramManager::link` doesn't use `m_ExportedSymbolImages`, so all of this makes sense.

> Thoughts?

Yes, I think we should disable all implicit linking (exception: the special bfloat16 device library) for RTC bundles, otherwise I think you could construct silly examples where one RTC bundle uses a function from another, or maybe even from a normal (non-RTC) device image.

Reply:
Doing it in a follow-up makes sense 👍